### PR TITLE
jsx documentation + example

### PR DESCRIPTION
**Pull request by nojvek**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -106,6 +106,8 @@ patch(newVnode, null)
 * [Animated reordering of elements](http://snabbdom.github.io/snabbdom/examples/reorder-animation/)
 * [Hero transitions](http://snabbdom.github.io/snabbdom/examples/hero/)
 * [SVG Carousel](http://snabbdom.github.io/snabbdom/examples/carousel-svg/)
+* [TSX Minesweeper](http://snabbdom.github.io/snabbdom/examples/tsx-minesweeper/)
```
Comment (nojvek) [inline README.md]: May be the example could be simpler: like a clock example

File: README.md [hunk 2]
```diff
@@ -580,6 +582,93 @@ The view function here is only an example. In practice thunks are only
...
+## JSX
+
+[JSX](https://facebook.github.io/jsx/) is a xml-like syntax extension to ecmascript.
+Instead of using `h(tag, props, children)` to define the virtual tree,
+snabbdom supports using jsx in babel or tsx in typescript.
+
+Built-in jsx function is a very thin and performant layer on top of `h`.
+Top level props are the initialized modules such as `class`, `attrs`, `props`, `on`, `style`.
+
+### Typescript
```
Comment (kuraga) [inline README.md]: `TypeScript`?

File: README.md [hunk 3]
```diff
@@ -580,6 +582,93 @@ The view function here is only an example. In practice thunks are only
 relevant if you are rendering a complicated view that takes
 significant computational time to generate.
 
+## JSX
+
+[JSX](https://facebook.github.io/jsx/) is a xml-like syntax extension to ecmascript.
```
Comment (kuraga) [inline README.md]: XML? ECMAScript?

File: README.md [hunk 4]
```diff
@@ -580,6 +582,93 @@ The view function here is only an example. In practice thunks are only
...
+Instead of using `h(tag, props, children)` to define the virtual tree,
+snabbdom supports using jsx in babel or tsx in typescript.
```
Comment (kuraga) [inline README.md]: Same here...

File: README.md (hunk 3 above)
Comment (nojvek) [reply README.md]: That's right. See https://facebook.github.io/jsx/

File: README.md (hunk 4 above)
Comment (nojvek) [reply README.md]: This comment feels a bit vague. What do you mean by `here` ?

I'd really appreciate more actionable PR comments in the format.

`{thing} is/has {issue} because {reason}, How about {suggestion} so it could be {ideal thing}?`

**Comment by nojvek** [comment]
@paldepind / @mightyiam want to give me a review pass ?

File: README.md (hunk 3 above)
Comment (kuraga) [reply README.md]: I meant case of letters. `xml` should be `XML`. `ecmascript` is `ECMAScript` indeed.

File: README.md (hunk 4 above)
Comment (kuraga) [reply README.md]: Cases of letters are wrong IMHO. `snabbdom` -> `Snabbdom`, `babel` -> `Babel`, `jsx` -> `JSX`, `tsx` -> `TSX`, `typescript` -> `TypeScript`

**Comment by nojvek** [comment]
kind ping @paldepind ^

File: examples/tsx-clock/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+      "outFile": "./build.js",
+      "preserveWatchOutput": true,
+      "removeComments": false,
+      "sourceMap": true,
+      "target": "ES6",
+      "lib": [
+          "dom",
+          "es2017",
+      ]
+  },
```
Comment (mightyiam) [inline examples/tsx-clock/tsconfig.json]: Consider using my https://github.com/mightyiam/tsconfigs.

File: jsx-global.d.ts [hunk 6]
```diff
@@ -1,4 +1,4 @@
-import {VNode, VNodeData} from './vnode';
+import {VNode, VNodeData} from './src/vnode';
```
Comment (mightyiam) [inline jsx-global.d.ts]: Why is this moved, please?

File: package.json [hunk 7]
```diff
@@ -91,6 +91,7 @@
     "src/hooks.ts",
     "src/htmldomapi.ts",
     "src/is.ts",
+    "src/jsx.ts",
```
Comment (mightyiam) [inline package.json]: Is this a fix? If so, an issue should be opened and this fixed in a separate PR, if you ask me. Same for the rest of the changes in this file.

**Comment by nojvek** [comment]
> I do not feel comfortable with having an built artifact checked in to version control. I don't like that. Could we avoid that, please? How about deploying to some service, instead? How about https://zeit.co/ ?

The rest of the examples had built files, so I followed convention. 

I do like that examples are hosted by github as flat files and are in the repo. It’s close to rest of other documentation and we don’t have to depend on another third party. 

I’ll make the other changes. 

Re JSX example: I wanted to have something neat and fun with function components. Not too complex but enough to show how a great performant app could be built. I was going to go with something like TodoMVC but this seemed simpler and more fun to look at. 

Happy to take proposals


Will open another PR to move jsx-globals.d.ts to repo root and keep this pure documentation.

**Comment by nojvek** [comment]
But that being said I’m debating with myself whether JSX even makes sense. I feel @paldepind was right. We aren’t changing the h function (except handling null) 

This feels like increasing surface area and could belong to another module, where the pace of improving it is much faster. I’m working with Typescript folks to land jsx fragment support and it feels supporting JSX could be an ongoing thing. 

If you’re okay @mightyjam, I think we should revert the existing JSX PR and only leave the simple null handing change in there. 

With the null change h function will still support jsx out of the box with children in array. So all is not lost.

**Comment by kuraga** [comment]
:+1: to ignore compiled example files.

**Comment by mightyiam** [comment]
Pending #497 .

**Comment by paldepind** [comment]
> Re JSX example: I wanted to have something neat and fun with function components. Not too complex but enough to show how a great performant app could be built. I was going to go with something like TodoMVC but this seemed simpler and more fun to look at.

I think that is fine and in line with the existing examples. They aren't completely stripped down but actually, somewhat, pleasant and fun.

I think having the build files is fine given that the other examples currently also have build files. One could argue that having prebuilt files makes it easier for someone to take a quick look at the example. But, one could also argue that having build files in git is annoying. In any case, I think that should be discussed in a different PR/issue and this PR can follow the current convention until then.

File: README.md [hunk 8]
```diff
@@ -163,7 +164,7 @@ var vnode = h('div', {style: {color: '#000'}}, [
 
 ### `snabbdom/tovnode`
 
-Converts a DOM node into a virtual node. Especially good for patching over an pre-existing, 
+Converts a DOM node into a virtual node. Especially good for patching over an pre-existing,
```
[the remark this answers is not in the document]
Comment (nojvek) [reply README.md]: editor (vscode) is simply stripping trailing whitespace on save. trailing whitespace doesn't seem very useful here does it? Has no visual change in README.md and seems like a good fix.

File: examples/tsx-clock/tsconfig.json (hunk 5 above)
Comment (nojvek) [reply examples/tsx-clock/tsconfig.json]: That's a cool repo, but I kinda wanted the example to have the raw base tscconfig.json without any extends redirections for simplicity. Plus that package isn't in package.json right?

File: README.md [hunk 9]
```diff
@@ -61,7 +61,6 @@ performance, small size and all the features listed below.
   * Powerful event listener functionality.
   * [Thunks](#thunks) to optimize the diff and patch process even further.
 * Third party features
-  * JSX support thanks to [snabbdom-pragma](https://github.com/Swizz/snabbdom-pragma).
```
Comment (nojvek) [inline README.md]: calling out JSX modules in JSX section

**Comment by nojvek** [comment]
@mightyiam ready for another pass. NOTE: it depends on merging #502 before though. When that lands I want to update `TSX Clock` example to use sel shorthand which makes things more readable.

**Comment by nojvek** [comment]
@mightyiam another review pass here please.

**Comment by nojvek** [comment]
kind ping @paldepind / @mightyiam ^

File: README.md [hunk 10]
```diff
@@ -611,6 +616,82 @@ The view function here is only an example. In practice thunks are only
...
 significant computational time to generate.
 
+## JSX
+
+[JSX](https://facebook.github.io/jsx/) is an XML-like syntax extension to JavaScript (ECMAScript).
+Instead of using `h(tag, props, children)` to define the virtual tree,
+Snabbdom supports using JSX with transpilers such as Babel and Typescript.
+
+Built-in `jsx` function is a very thin and performant layer on top of `h`.
+Top level props are the initialized [modules](#modules-documentation) such as `class`, `attrs`, `props`, `on`, `style`.
```
Comment (krail) [inline README.md]: What about `hooks`? Are they supported?

Comment (nojvek) [reply README.md]: yeah, it re-routes to underlying h, so everything is supported. I can add `hooks` to the docs.

Comment (krail) [reply README.md]: Thanks for clarification. It would be nice to add.

**Comment by nojvek** [comment]
@paldepind would also love a pass from you on this PR. 

Is there any chance you can guarantee us an SLA of minimum one review per week for PRs that need your approval to be merged and published.  

What I assumed would take a week max has almost spanned for 3 months now.

**Comment by nojvek** [comment]
This has been open for 6 months. I doubt @paldepind will ever come back to approve this. 6 months is ridiculous. Closing it.

**Comment by kuraga** [comment]
@nojvek , I'm sorry about this situation very much.

I saw you extracted this functionality into https://github.com/nojvek/snabbdom-jsx-lite. That's very cool!

Let's add this information to ReadMe, @nojvek , @mightyiam ?

**Comment by mightyiam** [comment]
But now that I have permission to make decisions freely, we can merge in JSX properly.

**Comment by nojvek** [comment]
Would love to merge in snabbdom. It’s mostly just one file. 

It’s cool that @mightyiam has full permissions now. That’s cool. 

I can open a PR to bring in the changes.

**Comment by mightyiam** [comment]
Please do, @nojvek. And feel free to contribute.

**Comment by mightyiam** [comment]
By 'contribute', I mean open pull requests. They will be reviewed.
